File: Editorial/65_Valid_Number/TypeScript/1_Follow_The_Rules.ts
function isNumber(s: string): boolean {
    let seenDigit: boolean = false;
    let seenExponent: boolean = false;
    let seenDot: boolean = false;
    for (let i = 0; i < s.length; i++) {
        let curr: string = s[i];
        if (!isNaN(parseInt(curr))) {
            seenDigit = true;
        } else if (curr == "+" || curr == "-") {
            if (i > 0 && s[i - 1] != "e" && s[i - 1] != "E") {
                return false;
            }
        } else if (curr == "e" || curr == "E") {
            if (seenExponent || !seenDigit) {
                return false;
            }
            seenExponent = true;
            seenDigit = false;
        } else if (curr == ".") {
            if (seenDot || seenExponent) {
                return false;
            }
            seenDot = true;
        } else {
            return false;
        }
    }
    return seenDigit;
}

File: Editorial/65_Valid_Number/C/1_Follow_The_Rules.c
bool isNumber(char *s) {
    bool seenDigit = false;
    bool seenExponent = false;
    bool seenDot = false;
    int len = strlen(s);
    for (int i = 0; i < len; i++) {
        char curr = s[i];
        if ('0' <= curr && curr <= '9') {
            seenDigit = true;
        } else if (curr == '+' || curr == '-') {
            if (i > 0 && s[i - 1] != 'e' && s[i - 1] != 'E') {
                return false;
            }
        } else if (curr == 'e' || curr == 'E') {
            if (seenExponent || !seenDigit) {
                return false;
            }
            seenExponent = true;
            seenDigit = false;
        } else if (curr == '.') {
            if (seenDot || seenExponent) {
                return false;
            }
            seenDot = true;
        } else {
            return false;
        }
    }
    return seenDigit;
}
